Add an option for printing syscall parameter names

xfgusta opened the pull request:
This is a partial implementation of #155. Using `-N` (mnemonic for "names") or `--arg-names` makes **strace** print the syscall parameter names. These names are taken from the corresponding libc function wrapper man pages. For example, `execve(2)` is declared as:

```c
int execve(const char *pathname, char *const _Nullable argv[], char *const _Nullable envp[]);
```

Given that, the output becomes:

```console
$ strace --arg-names -e trace=execve true
execve(pathname="/usr/bin/true", argv=["true"], envp=0x7ffc1fb49df8 /* 70 vars */) = 0
+++ exited with 0 +++
```

Argument names are available for most syscalls.

Here's how the output looks like with the changes:

```console
$ strace --arg-names -- /usr/bin/echo -n "Strauss"
execve(pathname="/usr/bin/echo", argv=["/usr/bin/echo", "-n", "Strauss"], envp=0x7fff060375b0 /* 70 vars */) = 0
brk(addr=NULL)                          = 0x55e8905c7000
mmap(addr=NULL, length=8192, prot=PROT_READ|PROT_WRITE, flags=MAP_PRIVATE|MAP_ANONYMOUS, fd=-1, offset=0) = 0x7f0c3c32a000
access(pathname="/etc/ld.so.preload", mode=R_OK) = -1 ENOENT (No such file or directory)
openat(dirfd=AT_FDCWD, pathname="/etc/ld.so.cache", flags=O_RDONLY|O_CLOEXEC) = 3
fstat(fd=3, statbuf={st_mode=S_IFREG|0644, st_size=95539, ...}) = 0
mmap(addr=NULL, length=95539, prot=PROT_READ, flags=MAP_PRIVATE, fd=3, offset=0) = 0x7f0c3c312000
close(fd=3)                             = 0
openat(dirfd=AT_FDCWD, pathname="/lib64/libc.so.6", flags=O_RDONLY|O_CLOEXEC) = 3
read(fd=3, buf="\177ELF\2\1\1\3\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\2607\0\0\0\0\0\0"..., count=832) = 832
pread64(fd=3, buf="\6\0\0\0\4\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0"..., count=784, offset=64) = 784
fstat(fd=3, statbuf={st_mode=S_IFREG|0755, st_size=2448320, ...}) = 0
pread64(fd=3, buf="\6\0\0\0\4\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0"..., count=784, offset=64) = 784
mmap(addr=NULL, length=2038872, prot=PROT_READ|PROT_EXEC, flags=MAP_PRIVATE|MAP_DENYWRITE, fd=3, offset=0) = 0x7f0c3c120000
mmap(addr=0x7f0c3c28f000, length=479232, prot=PROT_READ, flags=MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, fd=3, offset=0x16f000) = 0x7f0c3c28f000
mmap(addr=0x7f0c3c304000, length=24576, prot=PROT_READ|PROT_WRITE, flags=MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, fd=3, offset=0x1e3000) = 0x7f0c3c304000
mmap(addr=0x7f0c3c30a000, length=31832, prot=PROT_READ|PROT_WRITE, flags=MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, fd=-1, offset=0) = 0x7f0c3c30a000
close(fd=3)                             = 0
mmap(addr=NULL, length=12288, prot=PROT_READ|PROT_WRITE, flags=MAP_PRIVATE|MAP_ANONYMOUS, fd=-1, offset=0) = 0x7f0c3c11d000
arch_prctl(op=ARCH_SET_FS, addr=0x7f0c3c11d740) = 0
set_tid_address(tidptr=0x7f0c3c11da10)  = 652103
set_robust_list(addr=0x7f0c3c11da20, length=24) = 0
rseq(0x7f0c3c11d680, 0x20, 0, 0x53053053) = 0
mprotect(addr=0x7f0c3c304000, length=16384, prot=PROT_READ) = 0
mprotect(addr=0x55e8858fc000, length=4096, prot=PROT_READ) = 0
mprotect(addr=0x7f0c3c368000, length=8192, prot=PROT_READ) = 0
prlimit64(pid=0, resource=RLIMIT_STACK, new_limit=NULL, old_limit={rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
munmap(addr=0x7f0c3c312000, length=95539) = 0
getrandom(buf="\x48\xff\x30\x99\x00\x9d\xec\x03", buflen=8, flags=GRND_NONBLOCK) = 8
brk(addr=NULL)                          = 0x55e8905c7000
brk(addr=0x55e8905e8000)                = 0x55e8905e8000
openat(dirfd=AT_FDCWD, pathname="/usr/lib/locale/locale-archive", flags=O_RDONLY|O_CLOEXEC) = 3
fstat(fd=3, statbuf={st_mode=S_IFREG|0644, st_size=233242544, ...}) = 0
mmap(addr=NULL, length=233242544, prot=PROT_READ, flags=MAP_PRIVATE, fd=3, offset=0) = 0x7f0c2e200000
close(fd=3)                             = 0
fstat(fd=1, statbuf={st_mode=S_IFCHR|0600, st_rdev=makedev(0x88, 0x1), ...}) = 0
write(fd=1, buf="Strauss", count=7Strauss)     = 7
close(fd=1)                             = 0
close(fd=2)                             = 0
exit_group(status=0)                    = ?
+++ exited with 0 +++
```

Tests still need to be written, but I decided to open this PR to get feedback on the implementation, as well as some guidance on how the tests should be written.

ldv-alt reviewed (commented):
We have a precedence of printing syscall arguments with their names, see e.g.
`strace -qq --signal=none --trace=clone sh -c ':&'`.

The way it's implemented in clone.c via `tprints_arg_name_begin`/`tprint_arg_name_end` is a bit awkward, though, as `tprint_arg_name_end` doesn't seem to be that much useful.

So maybe all we need is `tprints_arg_name` and maybe also `tprints_arg_next_name` as a shortcut for `tprint_arg_next` followed by `tprints_arg_name`.

I'm not sure libc headers is a good reference point, function prototypes there do not serve as a documentation, I'd rather look at manual pages instead.

With regards to tests, the very least we need is to ensure the -N output is correct for all syscalls.

xfgusta commented:
> We have a precedence of printing syscall arguments with their names, see e.g. `strace -qq --signal=none --trace=clone sh -c ':&'`.
> 

That's interesting. I overlooked that. Since we already have this by default, maybe we could just extend it to other syscalls. There wouldn't be a need for a new option. I'm a bit worried since it's a big change to the output, but it does make sense if `clone` already uses it.

> The way it's implemented in clone.c via `tprints_arg_name_begin`/`tprint_arg_name_end` is a bit awkward, though, as `tprint_arg_name_end` doesn't seem to be that much useful.
> 
> So maybe all we need is `tprints_arg_name` and maybe also `tprints_arg_next_name` as a shortcut for `tprint_arg_next` followed by `tprints_arg_name`.
> 

That makes sense. In this case, adding `print_syscall_param` is unnecessary. I would replace it with `tprints_arg_name` or `tprints_arg_next_name` where appropriate.

> I'm not sure libc headers is a good reference point, function prototypes there do not serve as a documentation, I'd rather look at manual pages instead.
> 

Sorry, I didn't express myself precisely. The names actually came from the manual pages; I even used `man execve(2)` as an example. I'll update the description to make that clearer.

ldv-alt commented:
> > We have a precedence of printing syscall arguments with their names, see e.g. `strace -qq --signal=none --trace=clone sh -c ':&'`.
> 
> That's interesting. I overlooked that. Since we already have this by default, maybe we could just extend it to other syscalls. There wouldn't be a need for a new option. I'm a bit worried since it's a big change to the output, but it does make sense if `clone` already uses it.

No, I didn't suggest changing the default output, but rather making -N output compatible with the current output for `clone`.

xfgusta commented:
Hi,

I changed `--param-names` to `--arg-names`. Replaced `print_syscall_param` with `tprint_arg_name` and `tprint_arg_name_next`, implemented in `print_arg_name.c`. Tried doing this in `print_fields.h`, but couldn't access the `Nflag`; could probably do it using function-like macros. Let me know if that's preferable.

Moved the "Implement decoding of ..." commits into the first commit and removed the mention of "decoding" from the NEWS file. I tried making `tprint_arg_name`/`tprint_arg_name_next` compatible with `clone`, but this would probably require checking the syscall. For now, I decided to keep the current implementation with `tprints_arg_name_begin`.

The commit message was updated to reflect all these changes.

xfgusta commented:
>  Also, most of ioctls decoders remain unaffected by the change,

Most of them don't have man pages, so I guess "argp" could be printed. For now, I updated only those with man pages.

>  and maint/gen/codegen.c can also generate some code containing tprint_arg_next() calls.

Updated to use `tprints_arg_next_name`. It seems to only generate the src/gen/gen_hdio.c file. Looked into it and it seems fine; let me know if anything looks off.

Also addressed the changes you requested. Updated the commit message as well.

<details>
    <summary>Diff between the old and new commit messages</summary>

```diff
--- old-commit-message.txt      2025-07-19 15:44:58.744574418 -0300
+++ new-commit-message.txt      2025-07-19 15:26:55.706925332 -0300
@@ -1,14 +1,12 @@
 strace: add an option for printing syscall argument names
 
 * src/strace.c: Add -N/--arg-names option.
-* src/print_arg_name.c: New file.
-* src/Makefile.am (libstrace_a_SOURCES): Add exit.c, print_arg_name.c,
-set_tid_address.c.
-* src/defs.h (tprints_arg_name,
-tprints_arg_next_name): New function declarations.
+* src/Makefile.am (libstrace_a_SOURCES): Add exit.c, set_tid_address.c.
 * src/syscall_dummy.h (sys_alarm): Remove alias to printargs_u.
 (sys_exit, sys_pkey_free): Remove alias to printargs_d.
 (sys_set_tid_address): Remove alias to printargs.
+* src/print_fields.h (tprints_arg_name, tprints_arg_next_name): New functions.
+* maint/gen/codegen.c (generate_decoder): Use it.
 * src/access.c (decode_access, decode_faccessat): Use it.
 (SYS_FUNC(faccessat2)): Likewise.
 * src/affinity.c (SYS_FUNC(sched_setaffinity),
@@ -34,6 +32,7 @@
 (SYS_FUNC(close), SYS_FUNC(oldselect)): Likewise.
 * src/dirent.c (SYS_FUNC(readdir)): Use it.
 * src/dup.c (dup_123): Use it.
+* src/epoll_ioctl.c (epoll_ioctl): Use it.
 * src/epoll.c (SYS_FUNC(epoll_create), SYS_FUNC(epoll_create1),
 SYS_FUNC(epoll_ctl)): Use it.
 (epoll_wait_common, epoll_pwait_common): Likewise.
@@ -51,6 +50,8 @@
 * src/file_handle.c (SYS_FUNC(name_to_handle_at),
 SYS_FUNC(open_by_handle_at)): Use it.
 * src/flock.c (SYS_FUNC(flock)): Use it.
+* src/fs_0x94_ioctl.c (decode_file_dedupe_range, fs_0x94_ioctl): Use it.
+* src/fs_f_ioctl.c (fs_f_ioctl): Use it.
 * src/fsconfig.c (SYS_FUNC(fsconfig)): Use it.
 * src/fsmount.c (SYS_FUNC(fsmount)): Use it.
 * src/fsopen.c (SYS_FUNC(fsopen)): Use it.
@@ -66,6 +67,7 @@
 * src/getpid.c (SYS_FUNC(getpgid), SYS_FUNC(getsid),
 SYS_FUNC(setpgid)): Use it.
 * src/getrandom.c (SYS_FUNC(getrandom)): Use it.
+* src/hdio.c (print_hdio_getgeo, print_hdio_drive_cmd): Use it.
 * src/hostname.c (SYS_FUNC(sethostname), SYS_FUNC(gethostname)): Use it.
 * src/inotify.c (SYS_FUNC(inotify_add_watch), SYS_FUNC(inotify_rm_watch),
 SYS_FUNC(inotify_init1)): Use it.
@@ -89,6 +91,12 @@
 * src/ipc_shmctl.c (SYS_FUNC(shmctl)): Use it.
 * src/ipc.c (SYS_FUNC(ipc)): Use it.
 * src/kcmp.c (SYS_FUNC(kcmp)): Use it.
+* src/kd_ioctl.c (kiocsound, kd_mk_tone, kd_get_kb_type, kd_io, kd_set_mode,
+kd_get_mode, kd_screen_map, kd_uni_screen_map, kd_set_kbd_mode,
+kd_get_kbd_mode, kd_kbd_entry, kd_kbd_str_entry, kd_diacr, kd_diacr_uc,
+kd_keycode, kd_sigaccept, kd_kbdrep, kd_font, kd_kbmeta, kd_unimapclr,
+kd_cmap): Use it.
+* src/kd_mpers_ioctl.c (kd_unimap, kd_fontx, kd_font_op): Use it.
 * src/kexec.c (SYS_FUNC(kexec_load), SYS_FUNC(kexec_file_load)): Use it.
 * src/keyctl.c (SYS_FUNC(add_key), SYS_FUNC(request_key),
 SYS_FUNC(keyctl)): Use it.
@@ -144,6 +152,7 @@
 * src/pidfd_open.c (SYS_FUNC(pidfd_open)): Use it.
 * src/pkeys.c (SYS_FUNC(pkey_alloc), SYS_FUNC(pkey_free)): Use it.
 (SYS_FUNC(pkey_free)): New function.
+* src/nsfs.c (nsfs_ioctl): Use it.
 * src/poll.c (decode_poll_entering, do_poll, do_ppoll): Use it.
 * src/prctl.c (SYS_FUNC(prctl), SYS_FUNC(arch_prctl)): Use it.
 * src/process_vm.c (SYS_FUNC(process_vm_readv),
@@ -196,6 +205,7 @@
 * src/sysinfo.c (SYS_FUNC(sysinfo)): Use it.
 * src/syslog.c (SYS_FUNC(syslog)): Use it.
 * src/sysmips.c (SYS_FUNC(sysmips)): Use it.
+* src/term.c (term_ioctl): Use it.
 * src/time.c (SYS_FUNC(gettimeofday), SYS_FUNC(settimeofday),
 SYS_FUNC(osf_gettimeofday), SYS_FUNC(osf_settimeofday), SYS_FUNC(getitimer),
 SYS_FUNC(osf_getitimer), SYS_FUNC(setitimer), SYS_FUNC(osf_setitimer),
@@ -214,6 +224,7 @@
 * src/umount.c (SYS_FUNC(umount2)): Use it.
 * src/uname.c (SYS_FUNC(uname)): Use it.
 * src/userfaultfd.c (SYS_FUNC(userfaultfd)): Use it.
+(uffdio_ioctl): Likewise.
 * src/ustat.c (SYS_FUNC(ustat)): Use it.
 * src/utime.c (SYS_FUNC(utime)): Use it.
 * src/utimes.c (SYS_FUNC(utimes), SYS_FUNC(futimesat),

```

</details>

ldv-alt commented on maint/gen/codegen.c:
```diff
@@ -741,22 +741,24 @@ generate_decoder(FILE *out, struct syscall *syscall, bool is_variant,
 			generate_printer(out, syscall, arg.name, arg_val, true,
 					 arg.type, indent_level);
 
 			if (i < syscall->arg_count - 1) {
-				OUTSI("tprint_arg_next();\n");
+				OUTFI("tprints_arg_next_name(\"%s\");\n", arg.name);
```
`arg.name` here is the name of the argument that's already printed.
I suggest removing this `if` statement altogether.

ldv-alt commented on maint/gen/codegen.c:
```diff
@@ -741,22 +741,24 @@ generate_decoder(FILE *out, struct syscall *syscall, bool is_variant,
 		/* 0 out ptrs: print all args in sysenter */
 		for (size_t i = arg_offset; i < syscall->arg_count; i++) {
 			struct syscall_argument arg = syscall->args[i];
 			OUTFI("/* arg: %s (%s) */\n",
 			      arg.name, type_to_ctype(arg.type));
+			if (i == arg_offset) {
+				if (syscall->is_ioctl)
+					OUTFI("tprints_arg_next_name(\"%s\");\n", arg.name);
+				else
+					OUTFI("tprints_arg_name(\"%s\");\n", arg.name);
+			}
```
Wouldn't the following be better:
```c
if (i > arg_offset || syscall->is_ioctl)
        OUTFI("tprints_arg_next_name(\"%s\");\n", arg.name);
else
        OUTFI("tprints_arg_name(\"%s\");\n", arg.name);
```

ldv-alt commented on maint/gen/codegen.c:
```diff
@@ -782,13 +780,19 @@ generate_decoder(FILE *out, struct syscall *syscall, bool is_variant,
 			OUTFI("/* arg: %s (%s) */\n",
 			      arg.name, type_to_ctype(arg.type));
+			if (cur == arg_offset) {
+				if (syscall->is_ioctl)
+					OUTFI("tprints_arg_next_name(\"%s\");\n", arg.name);
+				else
+					OUTFI("tprints_arg_name(\"%s\");\n", arg.name);
+			}
 			CLEANUP_FREE char *arg_val =
 				get_syscall_arg_value(syscall, arg_index++);
 			generate_printer(out, syscall, arg.name, arg_val, true,
 					 arg.type, indent_level);
 
 			if (cur < syscall->arg_count - 1) {
-				OUTSI("tprint_arg_next();\n\n");
+				OUTFI("tprints_arg_next_name(\"%s\");\n\n", arg.name);
 			}
```
Likewise here.

ldv-alt commented on maint/gen/codegen.c:
```diff
@@ -830,7 +836,7 @@ generate_decoder(FILE *out, struct syscall *syscall, bool is_variant,
 					 false, arg.type, indent_level);
 
 			if (cur < syscall->arg_count - 1) {
-				OUTSI("tprint_arg_next();\n");
+				OUTFI("tprints_arg_next_name(\"%s\");\n", arg.name);
 			}
```
Likewise.

ldv-alt commented on maint/gen/codegen.c:
```diff
@@ -817,6 +821,8 @@ generate_decoder(FILE *out, struct syscall *syscall, bool is_variant,
 			struct syscall_argument arg = syscall->args[cur];
 			OUTFI("/* arg: %s (%s) */\n",
 			      arg.name, type_to_ctype(arg.type));
+			if (cur == arg_offset)
+				OUTFI("tprints_arg_next_name(\"%s\");\n", arg.name);
```
Likewise.

ldv-alt commented on src/epoll_ioctl.c:
```diff
@@ -40,14 +40,14 @@ epoll_ioctl(struct tcb *const tcp, const unsigned int code,
 
 	case EPIOCGPARAMS:
 		if (entering(tcp)) {
-			tprint_arg_next();
+			tprints_arg_next_name("argp");
 			return 0;
 		}
 		print_struct_epoll_params(tcp, arg);
```
I suggest moving this `tprints_arg_next_name` from the entering phase to the exiting, i.e. right before `print_struct_epoll_params`.

ldv-alt commented on src/hdio.c:
```diff
@@ -25,8 +25,7 @@ static int
 {
 	if (entering(tcp)) {
-		tprint_arg_next();
-
+		tprints_arg_next_name("argp");
 		return 0;
 	}
```
Likewise, I suggest moving this tprints_arg_next_name from the entering phase to the exiting.

ldv-alt commented on src/kd_ioctl.c:
```diff
@@ -172,7 +172,7 @@ kd_get_kb_type(struct tcb *const tcp, const kernel_ulong_t arg)
 	unsigned char val;
 
 	if (entering(tcp)) {
-		tprint_arg_next();
+		tprints_arg_next_name("argp");
 		return 0;
```
Likewise.

ldv-alt commented on src/kd_ioctl.c:
```diff
@@ -155,7 +155,7 @@ kd_leds(struct tcb *const tcp, const unsigned int code,
 	}
 
 	if (entering(tcp)) {
-		tprint_arg_next();
+		tprints_arg_next_name("argp");
 
 		if (get)
 			return 0;
```
Likewise.

ldv-alt commented on src/kd_ioctl.c:
```diff
@@ -216,7 +216,7 @@ kd_get_mode(struct tcb *const tcp, const kernel_ulong_t arg)
 	unsigned int val;
 
 	if (entering(tcp)) {
-		tprint_arg_next();
+		tprints_arg_next_name("argp");
 		return 0;
 	}
```
Likewise.

ldv-alt commented on src/kd_ioctl.c:
```diff
@@ -234,7 +234,7 @@ static int
 kd_screen_map(struct tcb *const tcp, const kernel_ulong_t arg, const bool get)
 {
 	if (entering(tcp)) {
-		tprint_arg_next();
+		tprints_arg_next_name("argp");
 
 		if (get)
 			return 0;
```
Likewise.

ldv-alt commented on src/kd_ioctl.c:
```diff
@@ -276,7 +276,7 @@ kd_uni_screen_map(struct tcb *const tcp, const kernel_ulong_t arg,
 	unsigned short elem;
 
 	if (entering(tcp)) {
-		tprint_arg_next();
+		tprints_arg_next_name("argp");
 
 		if (get)
 			return 0;
```
Likewise.

ldv-alt commented on src/kd_ioctl.c:
```diff
@@ -303,7 +303,7 @@ kd_get_kbd_mode(struct tcb *const tcp, const kernel_ulong_t arg)
 	unsigned int val;
 
 	if (entering(tcp)) {
-		tprint_arg_next();
+		tprints_arg_next_name("argp");
 		return 0;
 	}
```
Likewise.

ldv-alt commented on src/kd_ioctl.c:
```diff
@@ -480,7 +480,7 @@ kd_diacr(struct tcb *const tcp, const kernel_ulong_t arg, const bool get)
 	struct kbdiacr elem;
 
 	if (entering(tcp)) {
-		tprint_arg_next();
+		tprints_arg_next_name("argp");
 
 		if (get)
 			return 0;
```
Likewise.

ldv-alt commented on src/kd_ioctl.c:
```diff
@@ -528,7 +528,7 @@ kd_diacr_uc(struct tcb *const tcp, const kernel_ulong_t arg, const bool get)
 	struct_kbdiacruc elem;
 
 	if (entering(tcp)) {
-		tprint_arg_next();
+		tprints_arg_next_name("argp");
 
 		if (get)
 			return 0;
```
Likewise.

ldv-alt commented on src/kd_ioctl.c:
```diff
@@ -645,7 +645,7 @@ static int
 kd_font(struct tcb *const tcp, const kernel_ulong_t arg, const bool get)
 {
 	if (entering(tcp)) {
-		tprint_arg_next();
+		tprints_arg_next_name("argp");
 
 		if (get)
 			return 0;
```
Likewise.

ldv-alt commented on src/kd_ioctl.c:
```diff
@@ -670,7 +670,7 @@ kd_kbmeta(struct tcb *const tcp, const kernel_ulong_t arg, const bool get)
 	unsigned int val;
 
 	if (entering(tcp)) {
-		tprint_arg_next();
+		tprints_arg_next_name("argp");
 
 		if (get)
 			return 0;
```
Likewise.

ldv-alt commented on src/kd_ioctl.c:
```diff
@@ -717,7 +717,7 @@ static int
 kd_cmap(struct tcb *const tcp, const kernel_ulong_t arg, const bool get)
 {
 	if (entering(tcp)) {
-		tprint_arg_next();
+		tprints_arg_next_name("argp");
 
 		if (get)
 			return 0;
```
Likewise.

ldv-alt commented on src/ptp.c:
```diff
@@ -280,7 +280,7 @@ ptp_ioctl(struct tcb *const tcp, const unsigned int code,
 		CHECK_IOCTL_SIZE(PTP_SYS_OFFSET_PRECISE2, 64);
 
 		if (entering(tcp)) {
-			tprint_arg_next();
+			tprints_arg_next_name("argp");
 			return 0;
 		}
```
Let's move this `tprints_arg_next_name("argp");` invocation after the `entering` branch.

ldv-alt commented on src/tee.c:
```diff
@@ -162,7 +162,7 @@ tee_version(struct tcb *const tcp, const kernel_ulong_t arg)
 	struct tee_ioctl_version_data version;
 
 	if (entering(tcp)) {
-		tprint_arg_next();
+		tprints_arg_next_name("argp");
 		return 0;
 	}
```
Let's move this `tprints_arg_next_name("argp");` invocation after the `entering` branch.

ldv-alt commented on src/watchdog_ioctl.c:
```diff
@@ -26,7 +26,7 @@ watchdog_ioctl(struct tcb *const tcp, const unsigned int code,
 	switch (code) {
 	case WDIOC_GETSUPPORT:
 		if (entering(tcp)) {
-			tprint_arg_next();
+			tprints_arg_next_name("argp");
 			return 0;
 		}
```
Let's move this `tprints_arg_next_name("argp");` invocation after the `entering` branch.

ldv-alt reviewed (changes requested):
Good progress!

I've reviewed remaining tprint_arg_next() invocations, and found a few cases where syscall arguments aren't printed yet:
- in src/ipc_msg.c:SYS_FUNC(msgrcv), fetch_msgrcv_args() may print something in case of an error;
- in src/ipc_semctl.c:SYS_FUNC(semctl), umove_or_printaddr() may print something in case of an error;
- in src/keyctl.c, src/prctl, and src/s390.c, there are a quite a few cases when syscall arguments are printed without tprints_arg_name/tprints_arg_next_name;
- in src/util.c, printargs* functions do not print syscall argument names.

Also, in src/time.c:do_nanosleep, lets move the tprints_arg_next call to the exiting branch, and merge it with `tprints_arg_name("rem");`.

ldv-alt reviewed (approved):
I've addressed all remaining issues to get this merged before the release.
There is still a big question how to test this properly, but let's not block on this.

Thanks!

xfgusta commented:
Thanks for finishing the work and for the review! That's really nice!
